execution/exchange/duplicate_label.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package exchange
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/thanos-io/promql-engine/execution/model"
+	"github.com/thanos-io/promql-engine/extlabels"
+	"github.com/thanos-io/promql-engine/query"
+)
+
+type duplicateLabelCheckOperator struct {
+	model.OperatorTelemetry
+
+	once sync.Once
+	next model.VectorOperator
+
+	checkDuplicates func([]model.StepVector) error
+}
+
+func NewDuplicateLabelCheck(next model.VectorOperator, opts *query.Options) model.VectorOperator {
+	return &duplicateLabelCheckOperator{
+		OperatorTelemetry: model.NewTelemetry("[duplicateLabelCheck]", opts.EnableAnalysis),
+		next:              next,
+	}
+}
+
+func (d *duplicateLabelCheckOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := d.initOnce(ctx); err != nil {
+		return nil, err
+	}
+	start := time.Now()
+	defer func() { d.AddExecutionTimeTaken(time.Since(start)) }()
+
+	in, err := d.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		return nil, nil
+	}
+
+	if err := d.checkDuplicates(in); err != nil {
+		return nil, err
+	}
+	return in, nil
+}
+
+func (d *duplicateLabelCheckOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	return d.next.Series(ctx)
+}
+
+func (d *duplicateLabelCheckOperator) GetPool() *model.VectorPool {
+	return d.next.GetPool()
+}
+
+func (d *duplicateLabelCheckOperator) Explain() (me string, next []model.VectorOperator) {
+	return d.next.Explain()
+}
+
+func (d *duplicateLabelCheckOperator) initOnce(ctx context.Context) error {
+	var err error
+	d.once.Do(func() {
+		series, seriesErr := d.next.Series(ctx)
+		if err != nil {
+			err = seriesErr
+			return
+		}
+		d.checkDuplicates = checkDuplicatesFn(series)
+	})
+	return err
+}
+
+const (
+	bigPrime = 999686999
+)
+
+func checkDuplicatesFn(series []labels.Labels) func([]model.StepVector) error {
```

Comment:
Its just the simplest method i could come up with today, it might be stupid and can be discussed but is ultimately an implementation detail that we can change out as we want.